### ISSUE TITLE
questions on busy-waiting and CAS contention in compiler distribution and user documentation

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#11707]
### TL;DR

Concurrent-programming experts on the internet say that busy-waiting is a bad thing, and that CAS contention causes high overheads in contended settings. I looked at what we do in the Multicore runtime and what we encourage our users to do, and it seems that we are not always following best practices. This issue discusses what I understood and proposes some mostly-simple changes for discussion.

### Intro

I spent my Saturday night being nerd-sniped by descriptions of high-performance concurrent algorithms on the internet, in particular:

- [Building a Tiny Mutex](https://zig.news/kprotty/building-a-tiny-mutex-537k) (Protty, Zig news; also has a blog post on [production-level domain pool implementations](https://zig.news/kprotty/resource-efficient-thread-pools-with-zig-3291) if you are curious), which explains that CAS contention really hurts scalability, and

- [this email from Linus Torvalds](https://www.realworldtech.com/forum/?threadid=189711&curpostid=189759) in an ominous thread called "No nuances, just buggy code (was: related to Spinlock implementation and the Linux Scheduler)", which explains that "I'll spin for a short while before waiting" is generally a bad strategy.

(This is for context, you don't need to read these posts. The explanations below should be introductory-level and self-contained.)

Reading this made me realize that:

- Implementing synchronization primitives is hard and full of details I'm not aware of.
- Lock-free atomic programming, as is natural when mixing Domain and Atomic together,
  is extremely prone to reinventing bad/naive implementations of synchronization primitives.
- I've already jumped into naive, non-scalable concurrent code smells myself without noticing.
- We should ensure that our Multicore documentation does not encourage our users to do the same.

I think that it is useful to think about this now that we are about to unleash concurrent OCaml programming to our users using only bare-bone primitives.

Responsibilities are different in a language like Go that exposes an opinionated concurrency model from the start, and intentionally hides the details that users would need to shoot themselves in the foot by reimplementing bad locks or naive domain-local state. (See for example [Goroutines have no identity](https://notes.shichao.io/gopl/ch9/#goroutines-have-no-identity).)

## Spinning (busy-waiting)

(If you are an expert, you can skip this section and jump straight to "proposed changes" below.)

### Busy-looping

Busy-looping is the practice of looping until a concurrent operation succeeds, for example:

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/stdlib/atomic.mli#L131-L135

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/stdlib/domain.ml#L73-L76

These examples loop until the current domain manages to update the global state with its own computation.

### Busy-waiting (spinning)

A sub-category of busy-looping programs is busy-waiting or spinning, where we busy-loop over a condition that does not depend on the current domain/thread (as above: managing to update the state), but we wait on some other domain doing something.

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/manual/src/tutorials/parallelism.etex#L670-L673

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/runtime/domain.c#L995-L1000

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/runtime/domain.c#L925
```c
  msg = atomic_load_acq (&di->backup_thread_msg);
  while (msg != BT_TERMINATE) {
    switch (msg) {
      case BT_IN_BLOCKING_SECTION: /* ... */;
      case BT_ENTERING_OCAML: /* ... */
      default:
        cpu_relax();
        break;
    };
    msg = atomic_load_acq (&di->backup_thread_msg);
  }
```

The internet seems to say that while busy-looping makes sense in some cases (but see CAS contention below), busy-waiting is generally evil, because it works very very poorly when the thread that we are waiting for is not currently scheduled by the OS. (So: it looks very nice in tightly-controlled benchmarks, and wastes a lot of time and energy/battery on resource-contended systems.)

In the words of Linus Torvalds in [another post of the thread](https://www.realworldtech.com/forum/?threadid=189711&curpostid=189752):

> Good locking [...] needs to actively tell the system what you're yielding to (and optimally it would also tell the system about whether you care about fairness/latency or not - a lot of loads don't). [...] It's generally something like std::mutex, pthread_mutex_lock(), or perhaps a tuned thing that uses an OS-specific facility like "futex", where you do the nonblocking (and non-contended) case in user space using a shared memory location, but when you get contention you tell the OS what you're waiting for (and what you're waking up).

The recommendation here is to check whether waiting is necessary at all (`try_lock`), and then use an existing mutex implementation that will have a chance of telling the OS scheduler what thread it is waiting on.

### Bounded spinning

Some programs implement "bounded spinning": they try to busy-wait but only for a short, bounded amount of time, and do something less naive/wasteful (typically: they 'wait' using a mutex or other synchronization primitive).

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/runtime/domain.c#L350-L356

The [Linus post](https://www.realworldtech.com/forum/?threadid=189711&curpostid=189759) I cited at the beginning of the post argues that bounded spinning is still generally a bad idea, "unless you really know a lot about your workload".

> Even a very limited "let's busy-loop" can have serious disadvantages. Doing that "CPU yield" thing ("rep nop" on x86, "yield" instruction on arm64) can trigger the hypervisor to do things in virtualized environments, as can things like doing a cycle counter read to limit the time you wait etc. It might work fine on your machine, but it can cause unexpected behavior elsewhere.
>
> And if you only do a hard busy loop reading the lock value, you might be wasting time, energy and cause unnecessary cache contention. If the loop is limited to a small value, it might have very dubious value, and if you make it longer, the downsides can grow a lot. And the value tuning might be as much about your load as it is about the particular machine you are running on, etc, so you might again be tuning only for your development machine, not any real life situation.
>
> So to a first approximation, start off with that fast-case "trylock/tryunlock" model where you handle the case of absolutely no contention specially yourself. And then fall back to a sleeping real model like futexes immediately after that unless you really know a lot about your workload.

Technical note: the bounded-spinning code I have shown above is part of the core synchronization primitive of the Multicore OCaml GC, so it probably qualifies as "you really know about your workload". On the other hand, it does not actually implement the general recommendation above, because it does not follow the bounded-spinning by "directed" locking (that would inform the OS scheduler of who we are waiting for), it follows with a SPIN_WAIT loop that waits in an undirected manner by calling `usleep`.

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/runtime/domain.c#L359-L362

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/runtime/caml/platform.h#L83-L90

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/runtime/platform.c#L290-L303

(SPIN_WAIT starts by doing bounded spinning, which seems redundant with the explicit bounded-spinning earlier in the function, and then it uses an "exponential backoff" technique of sleeping for exponentially-increasing times -- 1.25 multiplier.)

## CAS contention and `cpu_relax`

(If you are an expert, you can skip this section and jump straight to "proposed changes" below.)

Another issue explained in [Building a Tiny Mutex](https://zig.news/kprotty/building-a-tiny-mutex-537k) is that it is bad to busy-loop on an atomic *write* to a shared instruction, including a CAS which apparently somehow counts as a write, at least on x86 systems. This is bad because those operations invalidate per-core caches of the main memory, with performance suffering on many-core systems if busy-waiting cores start repeatedly invalidating all other caches. So: CAS works fine in uncontended scenario, but they exhibit scalability problem in contended cases.

If you architect your whole application with compute-intensive domains that heavily push/pop to a concurrent stack using

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/stdlib/atomic.mli#L131-L135

you may observe poor scalability.

[Wikipedia says](https://en.wikipedia.org/wiki/Compare-and-swap):

> Instead of immediately retrying after a CAS operation fails, researchers have found that total system performance can be improved in multiprocessor systems—where many threads constantly update some particular shared variable—if threads that see their CAS fail use exponential backoff—in other words, wait a little before retrying the CAS.

This is where `cpu_relax()` enters the picture: it calls CPU instructions to encourage the current core to take it easy. (Note: this does not go through the OS and in particular it does not "yield" with the OS scheduler.) `cpu_relax()` is essential in busy-waiting code to avoid burning all CPU resources, but I understand that it also plays the role of "backoff" in the busy-looping examples above: in high-contention scenarios, implementing exponential backoff would be ideal, but at least `cpu_relax` reduces the contention somewhat.

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/manual/src/tutorials/parallelism.etex#L703-L706

## Proposed changes

Below I'm trying to turn these general considerations into concrete change proposals -- that I'm willing to try to implement, or of course leave to experts. Please keep in mind that I'm a complete beginner at concurrency, so the suggestions may be extremely naive. Just tell me when that happens :-)

### 'no user busy-waiting': Stop advertising busy-waiting

According to experts, busy-waiting is bad practice. Presumably we should not encourage our users to do it, and avoid documentation examples such as

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/manual/src/tutorials/parallelism.etex#L670-L673

I propose to send a PR to remove this example from the documentation -- possibly think of a different example.

### 'relaxed busy-looping': Ensure that all our busy-looping code uses `cpu_relax`

Some of our busy-looping code uses `cpu_relax`, and some does not. I propose to send PRs to systematically use `cpu_relax` when busy-waiting.

This is not clearly necessary in some scenarios (for example: code running on domain spawn is not performance-critical and its behavior in contended scenarios is of little importance). But users are going to look at the concurrent code in the OCaml distribution and imitate it, so we should try to inspire good defaults.

Question for experts: is there an easy way to do exponential backoff that we should consider instead of just `cpu_relax`?

### 'relax doc': Write some documentation for `Domain.cpu_relax ()` to discourage busy-waiting and explain its busy-looping usage

I propose to send a PR to provide some documentation for `cpu_relax`, which currently has very little:

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/stdlib/domain.mli#L73-L75

The documentation should discourage busy-waiting, and explain the interest of `cpu_relax` for busy-looping usage. (Or: we could instead recommend an exponential-backoff strategy, if we make one easy to use.)

Question for experts: `Domain.cpu_relax` does not only relax, it also has the effect of checking for stop-the-world requests. Does this change the way to use or document this function? (My beginner intuition is that OCaml code that busy-loops will have a poll inserted anyway, and the poll does the same request-servicing work. So having `Domain.cpu_relax` check for incoming requests does not actually matter much.)

### 'Atomic.modify': Provide easy-to-use Atomic functions that integrate backoff

We can document how users should write their busy-looping code, but it would be even better to provide clean stdlib abstractions to do it for them.

This is directly related to an in-flight PR:

Atomic.modify, Atomic.modify_get #10798  
https://github.com/ocaml/ocaml/pull/10798

(the PR does not currently include any backoff logic, now I'm convinced that it should definitely have some)

### 'runtime busy-waiting' Review busy-waiting-looking code in the OCaml runtime to study its behavior on contended system and experiment with locks instead

I cited some examples of runtime code that use busy-waiting; one in the main stop-the-world logic, and two in the backup-thread code.

Question for experts: are those uses of busy-waiting hurtful in presence of contention? Would it be possible to rely on OS-scheduler-aware synchronization primitives instead?

My beginner intuition is:

- Not easily for the stop-the-world logic; this code is performance-critical and we won't want to trade better high-contention behavior for worse common-case performance.
  
  (Stil: I wonder if a counting semaphore couldn't be used instead of SPIN_WAIT after the bounded-spinning case.)

- Probably yes for the backup-thread logic, which busy-waits in less performance-critical scenarios. (They all look related to initialization and termination of the domain and its backup thread.)  But the synchronization protocol there is already quite complex and it may be delicate to introduce yet other synchronization point without introducing subtle deadlocks.

**Comment by gasche** [comment]
Note: I am myself unsatisfied with the strong backseat-driving vibe in this issue. I'm asking critical questions on code written by people that know much more than I do about this topic, and have done much more work to make it work in the OCaml codebase. I hope it is clear that the intention is not to be negative about what is there, but to understand whether there indeed are issues or not, and possibly contribute to solving them. I'm sure that in any case the extra information provided by experts will be very useful to understand the codebase, and I can volunteer to send PRs with clarifying/documenting comments.

(In particular: above I proposed one possible explanation for why we actually want to `cpu_relax` when busy-looping (but not busy-waiting). This explanation may be completely wrong! But I would be interested to get a better explanation in this case, which could end up in the codebase itself.)

**Comment by gadmm** [comment]
This issue reminds me of the prototype multicore implementation of lazy that advocated busy-waiting with `cpu_relax`, which turned me sceptical because I also read a colorful quote of Linus warning about spinning in userland (this prototype was not merged eventually).

Probably spin-waiting should not be advocated without appropriate context in examples for users, and it would be nice if uses of `cpu_relax` in the runtime had ample comments (for instance I do not understand why `cpu_relax` is useful when there is already some unlocking-locking of mutex in the loop).

On the other hand this issue mixes several things. A CAS loop that does a unit of progress at every loop iteration is not the same as a spin-lock, and the paper cited by Wikipedia in your quote has nothing to do with inserting the `pause` instruction inside CAS loops.

**Comment by gasche** [comment]
(cc also @xavierleroy, @kayceesrk, @ctk21 and @Engil who know more about this and might want to comment.)

**Comment by kayceesrk** [comment]
Firstly, thanks for thinking about this and asking tough questions. Just recording my initial thoughts here. I will save the answer to specific issues for later. As @gadmm, there are many interconnected issues being discussed here, and it is possible that way may try to pry apart some of these to make forward progress.

## Manual

Firstly, regarding the manual, my primary intention to describe the low-level primitives is to inform the experts about what primitives are available. After all, we have been recommending that the users should use higher-level libraries such as domainslib and eio. For example, see

https://github.com/ocaml/ocaml/blob/762014ff2c7a15e3b694c5ca92005e5c93b8612c/manual/src/tutorials/parallelism.etex#L7-L11

If necessary, it may be useful to state again that these are low-level primitives intended to be used to build higher-level libraries, and not intended to be used directly. If we find that certain primitives are categorically better, then we should include them in the standard library and recommend them. For example, `Atomic.modify` with exponential backoff seems like a useful primitive to add. 

I should also say that whether certain primitives are "bad" depends on the complex combination of the workload + the machine. I have observed a large variance in performance between different machines. That said, In an uncontended setting, spin-waiting is not bad. And, replacing spin-wait loops with a `Mutex.t` is not always the right thing to do in the presence of user-level threads, as `Mutex.lock` blocks the entire domain.

## ocaml-multicore/lockfree

On the plus side, the multicore team has been building lockfree library (https://github.com/ocaml-multicore/lockfree) with the explicit goal of incorporating the state-of-the-art in the lockfree space. One thing I am keen to do is to ensure that the libraries included here are rigorously tested (model checked using the (very WIP) [dscheck](https://github.com/sadiqj/dscheck/)), and systematically and continuously benchmarked on a variety of architectures (as we do with Sandmark). 

Should this work be included as part of the OCaml standard library? The downsides are that 
1. The responsibility falls on the already stretched core team
2. The library can only evolve at the speed of the compiler release cycles

Either way, we shall continue building lockfree libraries and understanding how to make them work well with OCaml.

## Runtime system

Regarding the uses of spin locks in the runtime system, if the proposed changes are correct, simplify reasoning and/or improve performance, we should aim to incorporate them. 

I should say that we have spent a considerable amount of time optimising the backup threads implementation. Some of the relevant PRs are: https://github.com/ocaml-multicore/ocaml-multicore/pull/397, https://github.com/ocaml-multicore/ocaml-multicore/issues/387, https://github.com/ocaml-multicore/ocaml-multicore/pull/389. Any change will have to be benchmarked thoroughly. Thankfully, this should be easy with the sandmark nightly service (https://sandmark.tarides.com) which includes runs of the parallel benchmarks.

**Comment by kayceesrk** [comment]
> (for instance I do not understand why cpu_relax is useful when there is already some unlocking-locking of mutex in the loop).

`cpu_relax` doesn't make sense here unless perhaps the lock is a `try_lock`?

**Comment by DemiMarie** [comment]
Could waiting be integrated with effects somehow?  My understanding is that OCaml programs will be multiplexing many user-level “threads” onto a few OS-level threads.  Therefore, if the current task is blocked, the most obvious thing to do is to yield to another task on the same core, which is very fast.  GHC can probably provide some insight here, as they have been doing this for ages.

**Comment by DemiMarie** [comment]
Not stale until #12579 is merged.

**Comment by DemiMarie** [comment]
This is finally fixed I think.

**Comment by gasche** [comment]
#12579 has been merged so the busy-waiting logic in the runtime has been optimized and tuned, but I believe that some issues remain:

0. There still are various examples of busy-wait in the runtime (for example: one on backup-thread creation, one in `caml_free_domains`).
1. We still have a busy-wait loop as an example in the manual, which I still suspect is a bad idea.
2. We don't have support for exponential backoff in the standard library, and we don't offer another waiting abstraction either (besides `Mutex`, which is not concurrency-approach-agnostic).
